### ISSUE TITLE
Document the delay introduced while creating new files from templates

### ISSUE THREAD

**Issue by danielo515** [SilentVoid13/Templater#1180]
Found this:

https://github.com/SilentVoid13/Templater/blob/487805b5ad1fd7fbc145040ed82b4c41fc2c48e2/src/core/Templater.ts#L397-L400

It frustrated me a bit because I was building a script, and until I remembered this little detail I was not being able to modify the newly created file with other methods. I think this is the root cause of my problem, but let me know if I'm wrong.
Here is the context: 
https://forum.obsidian.md/t/why-i-need-to-wait-a-random-amount-of-time-after-creating-a-file-with-teplater/66066

**Comment by Zachatoo** [comment]
I don't believe that 300ms delay is the issue. More likely, it's the fact that you're trying to access the cache of a file that was just barely created.
```js
const newFile = await tp.file.create_new(template, file_name, true, parent_path);
// Get the tags from the newly ceated file, in case the template had them
const fileMetadata = await app.metadataCache.getFileCache(newFile);
```

When creating a file in Obsidian, Obsidian takes a short amount of time to update it's metadata cache before it's cache is available. You are immediately requesting the cache after creating the file. You have to wait until the metadata cache has resolved. This is an Obsidian thing, not a Templater thing.

You could listen for the [changed](https://github.com/obsidianmd/obsidian-api/blob/5178e27f0eb12741c79914f52146de8ff827e390/obsidian.d.ts#L2537C6-L2537C6) event and then try to get the new file cache if you really don't want to have a delay in your code.

**Comment by danielo515** [comment]
Thanks for your answer @Zachatoo. 
I'm not saying that what you said doesn't make sense, it makes complete sense.
However, if that is the case, then why is my code working knowing that my delay happens AFTER the request to getFileCache has resolved already?

```
    const newFile = await tp.file.create_new(template, file_name, true, parent_path);
    // Get the tags from the newly ceated file, in case the template had them
    const fileMetadata = await app.metadataCache.getFileCache(newFile);
    // Concat the tags from the parent with the tags from the new file/template
    const newTags = tp.frontmatter.tags.concat(fileMetadata.frontmatter.tags);
    console.info({ newTags, newFile })
    await sleep(300)
    // use the metadata-menu plugin to set the tags on the new file
    await metadataMenu.postValues(newFile, [{ name: 'tags', payload: { value: newTags.join(',') } }]);
```

The reason I think it was because of templater delay is because (I think) the file was being overwritten by templater. I need to debug this to confirm again.

**Comment by danielo515** [comment]
https://github.com/SilentVoid13/Templater/assets/2270425/73a072f1-8763-4cf3-9407-87f8a03d6a15

In this video you can clearly see how the file is created, populated with the values from the selected template at 00:06, then it is modified correctly for half a second (right before 00:07) and then overwritten again with the values of the template. I suggest to play the video at 0.75 speed to see it more clearly.

**Comment by Zachatoo** [comment]
I took parts of your script and updated them and they work great for me, maybe it's an issue with metadataMenu? Either way, doesn't seem related to this Templater delay from my testing so far.

create-file template
```js
<%*  
const parent_path = tp.config.active_file.parent;
const newFilename = tp.file.title + "-" + await tp.system.prompt("New file name");
const templateFile = tp.file.find_tfile("template-with-tags");
const newFile = await tp.file.create_new(templateFile, newFilename, true, parent_path);
await app.fileManager.processFrontMatter(newFile, (frontmatter) => {
  frontmatter.tags.push(...tp.frontmatter.tags);
});
-%>
```

template-with-tags template
```md
---
tags:
  - tag-from-template
---
```

https://github.com/SilentVoid13/Templater/assets/6936914/2a73f13f-551c-49a0-a0e4-6cb696887f13

**Comment by danielo515** [comment]
> 

It's interesting that we both ended trying to use processFrontmatter to see if it makes a difference.
In my experience, it seems to be more reliable and easier to use than going to a third party plugin.
However, the problem is still there.
Please nothe your approach differs from mine in two key parts:

1. You are just putting all the code in the template, while I am using a template that calls a user function. Not sure if this makes a difference, but it is worth mentioning
2. The template you are using is very simple, and it may not run into racing conditions like mine does. Try something like this:
```
<%*
  const title = await tp.user.ensure_title(tp);
  tR += "---" 
%>
tags:  <%tp.file.cursor(1)%>
aliases:  <%tp.file.cursor(2)%>
title:  <%* tR += title %>
created: <% tp.date.now() %>
<%* tR += "---" -%>

<%tp.file.cursor(3)%>
```

This is the content of ensure title:

```js
module.exports = async function (tp) {
    let title = tp.file.title
    if (title.startsWith("Untitled")) {
        title = await tp.system.prompt("Title");
        await tp.file.rename(title);
    }
    return title;
}
```

**Comment by Zachatoo** [comment]
I must be missing a step or something, that code you've provided works without any issues for me. Are we doing the same steps?

https://github.com/SilentVoid13/Templater/assets/6936914/d7fa1960-bac8-41ba-a598-ac325be1dc65

**Comment by danielo515** [comment]
> Are we doing the same steps?

No we are not 😄 
The flow is:

1. simple templater script to trigger an user command (subtopic.md): `<%* tp.user.subtopic(tp) _%>`
2. The user script does all we mentioned: 
```
const ensureArray = value => {
    if (Array.isArray(value)) return value;
    switch (typeof value) {
        case 'string':
            return value.split(',');
        default:
            return [value];
    }
}
module.exports = async function (tp) {
    if (!tp.config.active_file) {
        new tp.obsidian.Notice('This template must be executed from an file to act as a parent.')
        return;
    }
    let parent_title = tp.file.title
    const parent_path = tp.config.active_file.parent;
    const parent_tags = ensureArray(tp.frontmatter.tags);
    let subtopic = await tp.system.prompt("Subtopic");
    const file_name = `${tp.user.sanitize(parent_title)}__${tp.user.sanitize(subtopic)}`
    const title = `${parent_title}: ${subtopic}`
    const templates = app.vault.getFiles().filter(f => f.path.startsWith('templates'));
    const template = await tp.system.suggester(t => t.basename, templates);
    const newFile = await tp.file.create_new(template, file_name, true, parent_path);
    // Get the tags from the newly ceated file, in case the template had them
    const fileMetadata = await app.metadataCache.getFileCache(newFile);
    // Concat the tags from the parent with the tags from the new file/template
    const newTags = parent_tags.concat(fileMetadata.frontmatter?.tags || []);
    // Some random delay because templater is writing the template twice
    await sleep(300)
    await app.fileManager.processFrontMatter(newFile, (frontmatter) => {
        frontmatter.title = title;
        frontmatter.parent = parent_title;
        frontmatter.tags = newTags;
    })
    await app.vault.append(newFile, `Parent:: [[${parent_title}]]`)
}
```
3. On the user script, then you pick the template I've posted earlier, the one that ensures the tile
4. Then things get into race conditions. I suspect that my main problem could a combination of issues. Sometimes it could be real race conditions in the templates mentioned, some other times it could be folder templates misbehaving and overwriting the newly created file with the folder template, although I think it should only do if the file is empty. Gonna ensure the folder template is not interfering

**Comment by Zachatoo** [comment]
I still must be missing something. I have taken your templates and user scripts and they appear to work fine. The only difference I made was you didn't provide me with the `tp.user.sanitize` user script so I removed that part, though I assume that's a fairly quick synchronous method. Works for me with or without the delay. The video below is without the delay.

https://github.com/SilentVoid13/Templater/assets/6936914/3acd4859-f18a-4b9b-b19c-c3bf09a30bb7

subtopic.js
```js
const ensureArray = (value) => {
  if (Array.isArray(value)) return value;
  switch (typeof value) {
    case "string":
      return value.split(",");
    default:
      return [value];
  }
};
module.exports = async function (tp) {
  if (!tp.config.active_file) {
    new tp.obsidian.Notice(
      "This template must be executed from an file to act as a parent."
    );
    return;
  }
  let parent_title = tp.file.title;
  const parent_path = tp.config.active_file.parent;
  const parent_tags = ensureArray(tp.frontmatter.tags);
  let subtopic = await tp.system.prompt("Subtopic");
  const file_name = `${parent_title}__${subtopic}`; // I REMOVED THE tp.user.sanitize() HERE
  const title = `${parent_title}: ${subtopic}`;
  const templates = app.vault
    .getFiles()
    .filter((f) => f.path.startsWith("templates"));
  const template = await tp.system.suggester((t) => t.basename, templates);
  const newFile = await tp.file.create_new(
    template,
    file_name,
    true,
    parent_path
  );
  // Get the tags from the newly ceated file, in case the template had them
  const fileMetadata = await app.metadataCache.getFileCache(newFile);
  // Concat the tags from the parent with the tags from the new file/template
  const newTags = parent_tags.concat(fileMetadata.frontmatter?.tags || []);
  // Some random delay because templater is writing the template twice
  // await sleep(300); // I COMMENTED OUT THE DELAY HERE
  await app.fileManager.processFrontMatter(newFile, (frontmatter) => {
    frontmatter.title = title;
    frontmatter.parent = parent_title;
    frontmatter.tags = newTags;
  });
  await app.vault.append(newFile, `Parent:: [[${parent_title}]]`);
};
```

ensure_title.js
```js
module.exports = async function (tp) {
  let title = tp.file.title;
  if (title.startsWith("Untitled")) {
    title = await tp.system.prompt("Title");
    await tp.file.rename(title);
  }
  return title;
};
```

subtopic.md
```js
<%* tp.user.subtopic(tp) _%>
```

daniello515.md
```js
<%*
  const title = await tp.user.ensure_title(tp);
  tR += "---" 
%>
tags:  <%tp.file.cursor(1)%>
aliases:  <%tp.file.cursor(2)%>
title:  <%* tR += title %>
created: <% tp.date.now() %>
<%* tR += "---" -%>

<%tp.file.cursor(3)%>
```

**Comment by danielo515** [comment]
> though I assume that's a fairly quick synchronous method.

Yes, that is correct.

Turns out the problem right now (who knows if it always was) is the default template. I just deactivated the default template and now I don't experience any problems either. My understanding is that the delay was not it making to work, but it was giving the default template enough time (most of the times) to overwrite the created file, and then the rest of my script finished, giving me the false impression that I was succeeding.
The things that trigger the default template are a bit confusing, and unclear. I will love to have a list of which events trigger the default root template, and which ones don't.

Thanks for all your help and patience @Zachatoo
